File: libpretixsync/src/test/java/eu/pretix/libpretixsync/db/QuestionAnswerValidationTest.java
package eu.pretix.libpretixsync.db;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.fail;

import eu.pretix.libpretixsync.check.QuestionType;
import eu.pretix.libpretixsync.db.AbstractQuestion;
import eu.pretix.libpretixsync.db.Question;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;

import javax.xml.bind.ValidationException;
import java.util.Arrays;
import java.util.Collection;


@RunWith(Parameterized.class)
public class QuestionAnswerValidationTest {

    @Parameterized.Parameters
    public static Collection<Object[]> data() {
        return Arrays.asList(new Object[][]{
                {QuestionType.S, "a", "a"},
                {QuestionType.T, "b", "b"},
                {QuestionType.S, "", ""},
                {QuestionType.N, "3", "3"},
                {QuestionType.N, "2.56", "2.56"},
                {QuestionType.N, "abc", null},
                {QuestionType.B, "True", "True"},
                {QuestionType.B, "true", "True"},
                {QuestionType.B, "False", "False"},
                {QuestionType.B, "false", "False"},
                {QuestionType.B, "0", "False"},
                {QuestionType.B, "", "False"},
                {QuestionType.C, "3", "3"},
                {QuestionType.C, "4", null},
                {QuestionType.C, "A", null},
                {QuestionType.M, "3", "3"},
                {QuestionType.M, "12", "12"},
                {QuestionType.M, "3,12", "3,12"},
                {QuestionType.M, "3,12,6", null},
                {QuestionType.M, "6", null},
                {QuestionType.D, "2018-01-19", "2018-01-19"},
                {QuestionType.D, "2016-02-29", "2016-02-29"},
                {QuestionType.D, "2017-02-29", null},
                {QuestionType.D, "fooobar", null},
                {QuestionType.H, "12:20", "12:20"},
                {QuestionType.H, "25:30", null},
                {QuestionType.H, "Foo", null},
                {QuestionType.W, "2018-01-19T12:20", "2018-01-19T12:20"},
                {QuestionType.W, "2016-02-29T14:30", "2016-02-29T14:30"},
                {QuestionType.W, "2016-02-29T25:59", null},
                {QuestionType.W, "2017-02-01", null},
                {QuestionType.W, "fooobar", null},
                // TODO: Date, time, datetime
        });
    }

    private QuestionType questionType;
    private String input;
    private String expected;

    public QuestionAnswerValidationTest(QuestionType questionType, String input, String expected) {
        this.questionType = questionType;
        this.input = input;
        this.expected = expected;
    }

    @Test
    public void test() {
        Question q = new Question();
        q.setType(questionType);

        QuestionOption qo = new QuestionOption();
        qo.setServer_id(3L);
        qo.setValue("A");
        q.getOptions().add(qo);
        qo = new QuestionOption();
        qo.setServer_id(12L);
        qo.setValue("B");
        q.getOptions().add(qo);

        if (expected == null) {
            try {
                q.clean_answer(input);
                fail("Expected an ValidationException to be thrown");
            } catch (Question.ValidationException e) {
            }
        } else {
            try {
                assertEquals(q.clean_answer(input), expected);
            } catch (AbstractQuestion.ValidationException e) {
                fail(e.getMessage());
            }
        }
    }
}

File: libpretixsync/src/test/java/eu/pretix/libpretixsync/db/QuestionAnswerValidationTest.kt
package eu.pretix.libpretixsync.db

import eu.pretix.libpretixsync.check.QuestionType
import org.junit.Assert.assertEquals
import org.junit.Assert.fail
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.Parameterized
import eu.pretix.libpretixsync.models.Question as QuestionModel


@RunWith(Parameterized::class)
class QuestionAnswerValidationTest(private val questionType: QuestionType, private val input: String, private val expected: String?) {

    @Test
    fun test() {
        val q = QuestionModel(
            type = questionType,
            required = false,
            question = "Test",
            identifier = "ABTBAB8S",
            askDuringCheckIn = false,
            options = listOf(
                QuestionOption(3L, 2, "AAA1", "A"),
                QuestionOption(12L, 1, "BBB2", "B"),
            ),
            id = 1L,
            serverId = 1L,
            eventSlug = "",
            position = 0L,
        )

        if (expected == null) {
            try {
                q.clean_answer(input, q.options, false)
                fail("Expected an ValidationException to be thrown")
            } catch (e: QuestionLike.ValidationException) {
            }

        } else {
            try {
                assertEquals(q.clean_answer(input, q.options, false), expected)
            } catch (e: QuestionLike.ValidationException) {
                fail(e.message)
            }

        }
    }

    companion object {

        @JvmStatic
        @Parameterized.Parameters
        fun data() = listOf(
                arrayOf(QuestionType.S, "a", "a"),
                arrayOf(QuestionType.T, "b", "b"),
                arrayOf(QuestionType.S, "", ""),
                arrayOf(QuestionType.N, "3", "3"),
                arrayOf(QuestionType.N, "2.56", "2.56"),
                arrayOf(QuestionType.N, "abc", null),
                arrayOf(QuestionType.B, "True", "True"),
                arrayOf(QuestionType.B, "true", "True"),
                arrayOf(QuestionType.B, "False", "False"),
                arrayOf(QuestionType.B, "false", "False"),
                arrayOf(QuestionType.B, "0", "False"),
                arrayOf(QuestionType.B, "", "False"),
                arrayOf(QuestionType.C, "3", "3"),
                arrayOf(QuestionType.C, "4", null),
                arrayOf(QuestionType.C, "A", null),
                arrayOf(QuestionType.M, "3", "3"),
                arrayOf(QuestionType.M, "12", "12"),
                arrayOf(QuestionType.M, "3,12", "3,12"),
                arrayOf(QuestionType.M, "3,12,6", null),
                arrayOf(QuestionType.M, "6", null),
                arrayOf(QuestionType.D, "2018-01-19", "2018-01-19"),
                arrayOf(QuestionType.D, "2016-02-29", "2016-02-29"),
                arrayOf(QuestionType.D, "2017-02-29", null),
                arrayOf(QuestionType.D, "fooobar", null),
                arrayOf(QuestionType.H, "12:20", "12:20"),
                arrayOf(QuestionType.H, "25:30", null),
                arrayOf(QuestionType.H, "Foo", null),
                arrayOf(QuestionType.W, "2018-01-19T12:20", "2018-01-19T12:20"),
                arrayOf(QuestionType.W, "2016-02-29T14:30", "2016-02-29T14:30"),
                arrayOf(QuestionType.W, "2016-02-29T25:59", null),
                arrayOf(QuestionType.W, "2017-02-01", null),
                arrayOf(QuestionType.W, "fooobar", null),
                arrayOf(QuestionType.EMAIL, "foobar@example.org", "foobar@example.org"),
                arrayOf(QuestionType.EMAIL, "foobar", null),
                arrayOf(QuestionType.EMAIL, "foobar@example.com.k", null)
        )
    }
}